# Task 4.4 - Practice work

# Importing libraries

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

# Importing data

In [2]:
# Importing data file - orders.csv
df_ords = pd.read_csv(r'C:\Users\srira\OneDrive\Documents\Instacart Basket Analysis\Data\Original Data\orders.csv', index_col = False)

In [3]:
# Importing data file - products.csv
df_prds = pd.read_csv(r'C:\Users\srira\OneDrive\Documents\Instacart Basket Analysis\Data\Original Data\products.csv', index_col = False)

# Data Wrangling

In [4]:
# Dropping eval_set from orders.csv
df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [5]:
df_ords = df_ords.drop(columns = ['eval_set'])

In [6]:
# Checking for null values in the days_since_prior_order variable
df_ords['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

In [7]:
# Renaming column order_dow as orders_days_of_week in orders.csv file
df_ords.rename(columns = {'order_dow' : 'orders_days_of_week'}, inplace = True)

In [8]:
df_ords.head()

order_id  user_id  order_number  orders_days_of_week  order_hour_of_day  \
0   2539329        1             1                    2                  8   
1   2398795        1             2                    3                  7   
2    473747        1             3                    3                 12   
3   2254736        1             4                    4                  7   
4    431534        1             5                    4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [9]:
# Changing the data type of the column order_id to 'str'
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [10]:
df_ords['order_id'].dtype

dtype('O')

# Transposing Data

In [12]:
path = r'C:\Users\srira\OneDrive\Documents\Instacart Basket Analysis'

In [14]:
# Importing data file departments.csv
df_dep = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'departments.csv'), index_col = False)

In [15]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [16]:
# Transposing df_dep
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [17]:
df_dep_t = df_dep.T

In [18]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [19]:
# Adding index to the dataframe
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [20]:
# Take the first row of the df_dep_t for the header
new_header = df_dep_t.iloc[0]

In [21]:
new_header

0    department
Name: department_id, dtype: object

In [23]:
# Take the data under the header row for a new df
df_dep_t_new = df_dep_t[1:]

In [24]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [25]:
# Set the header row as the df header
df_dep_t_new.columns = new_header

In [26]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

# Data Dictionary

In [28]:
# Turning the department dataframe into a dictionary
data_dict = df_dep_t_new.to_dict('index')

In [29]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [31]:
# First five entries of the products.csv dataframe
df_prds.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [33]:
print(data_dict.get('19'))

{'department': 'snacks'}


# Subsetting

In [34]:
# Data only from the snacks department
df_prds['department_id'] == 19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [35]:
# Products within df_prds with a department_id of 19 are returned
df_prds[df_prds['department_id'] == 19]

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [36]:
df_snacks = df_prds[df_prds['department_id'] == 19]

In [38]:
# Snacks department details
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

# Task 4.4 - Exercise

# Question 2: Finding an identifier variable and change data type

In [39]:
# Identifier variable in the df_ords dataframe
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [40]:
df_ords['user_id'].dtype

dtype('O')

# Question 3: Look for a variable in the df_ords dataframe with an unintuitive name and change its name without overwriting the dataframe.

In [42]:
# Looking into all the column names to identify a variable.
df_ords.columns


Index(['order_id', 'user_id', 'order_number', 'orders_days_of_week',
       'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')

In [44]:
# changing name of column without overwriting existing column name
df_ords.rename(columns = {'order_number' : 'number_of_orders'}, inplace = False)

order_id user_id  number_of_orders  orders_days_of_week  \
0        2539329       1                 1                    2   
1        2398795       1                 2                    3   
2         473747       1                 3                    3   
3        2254736       1                 4                    4   
4         431534       1                 5                    4   
...          ...     ...               ...                  ...   
3421078  2266710  206209                10                    5   
3421079  1854736  206209                11                    4   
3421080   626363  206209                12                    1   
3421081  2977660  206209                13                    1   
3421082   272231  206209                14                    6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

In [45]:
# variable name not overwritten
df_ords.head()

order_id user_id  order_number  orders_days_of_week  order_hour_of_day  \
0  2539329       1             1                    2                  8   
1  2398795       1             2                    3                  7   
2   473747       1             3                    3                 12   
3  2254736       1             4                    4                  7   
4   431534       1             5                    4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [46]:
# Renamed variable and saved in a new dataframe
df_orders = df_ords.rename(columns = {'order_number' : 'number_of_orders'}, inplace = False)

In [47]:
# New Dataframe
df_orders.head()

order_id user_id  number_of_orders  orders_days_of_week  order_hour_of_day  \
0  2539329       1                 1                    2                  8   
1  2398795       1                 2                    3                  7   
2   473747       1                 3                    3                 12   
3  2254736       1                 4                    4                  7   
4   431534       1                 5                    4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

# Question 4: Busiest hour for placing orders

In [48]:
# Count of the variable - order_hour_of_day
df_ords['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

# The busiest hours for placing orders is from 9AM to 4PM.

# Question 5:


In [ ]:
# Determining the meaning behind the value of 4 in the department_id column within the df_prds using a data dictionary.

In [92]:
# Importing data file - department.csv
df_dep = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'departments.csv'), index_col = False)

In [93]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [94]:
# Transposing df_dep
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [95]:
df_dep_t = df_dep.T

In [96]:
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [97]:
# Adding index to the dataframe
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [98]:
# Take the first row of the df_dep_t for the header
new_header = df_dep_t.iloc[0]

In [99]:
new_header

0    department
Name: department_id, dtype: object

In [100]:
# Take the data under the header row for a new df
df_dep_t_new = df_dep_t[1:]

In [101]:
# Set the header row as the df header
df_dep_t_new.columns = new_header

In [102]:
# Department dataframe
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [103]:
# Changing the department dataframe into a dictionary
data_dict = df_dep_t_new.to_dict('index')

In [104]:
# Data Dictionary
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [105]:
# Meaning of department_id 4
print(data_dict.get('4'))

{'department': 'produce'}


# The meaning behind the value of 4 in the department_id is the "produce" department.

# Question 6: More information about breakfast item sales

In [56]:
# Finding the department_id for breakfast
df_dep_t_new[df_dep_t_new['department'] == 'breakfast']

department_id department
14             breakfast

In [57]:
# More information on breakfast item sales
df_breakfast = df_prds[df_prds['department_id'] == 14]

In [59]:
df_breakfast.reset_index()

index  product_id                                       product_name  \
0        27          28                                  Wheat Chex Cereal   
1        33          34                                                NaN   
2        67          68                            Pancake Mix, Buttermilk   
3        89          90                                       Smorz Cereal   
4       210         211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...     ...         ...                                                ...   
1111  49330       49326                            Cereal Variety Fun Pack   
1112  49395       49391            Light and Fluffy Buttermilk Pancake Mix   
1113  49547       49543                          Chocolate Cheerios Cereal   
1114  49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
1115  49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

      aisle_id  department_id  prices  
0          121             14    10.1  
1          121             14    12.2  
2          130             14    13.7  
3          121             14     3.9  
4          130             14     3.6  
...        ...            ...     ...  
1111       121             14     9.1  
1112       130             14     2.0  
1113       121             14    10.8  
1114       130             14    14.2  
1115        57             14    10.2  

[1116 rows x 6 columns]

# Question 7: Details about products for dinner

In [61]:
# Finding department_id for alcohol, deli, beverages and meat/seafood
df_dep_t_new[df_dep_t_new['department'].isin(['alcohol', 'deli', 'beverages', 'meat seafood'])]

department_id    department
5                   alcohol
7                 beverages
12             meat seafood
20                     deli

In [62]:
# More information on dinner items
df_dinner = df_prds[df_prds['department_id'].isin([5, 7, 12, 20])]

In [63]:
df_dinner.reset_index()

index  product_id                                    product_name  \
0         2           3            Robust Golden Unsweetened Oolong Tea   
1         6           7                  Pure Coconut Water With Orange   
2         9          10  Sparkling Orange Juice & Prickly Pear Beverage   
3        10          11                               Peach Mango Juice   
4        16          17                               Rendered Duck Fat   
...     ...         ...                                             ...   
7645  49676       49672                          Cafe Mocha K-Cup Packs   
7646  49679       49675             Cinnamon Dolce Keurig Brewed K Cups   
7647  49680       49676                          Ultra Red Energy Drink   
7648  49686       49682                              California Limeade   
7649  49688       49684       Vodka, Triple Distilled, Twist of Vanilla   

      aisle_id  department_id  prices  
0           94              7     4.5  
1           98              7     4.4  
2          115              7     8.4  
3           31              7     2.8  
4           35             12    17.1  
...        ...            ...     ...  
7645        26              7     6.5  
7646        26              7    14.0  
7647        64              7    14.5  
7648        98              7     4.3  
7649       124              5     5.3  

[7650 rows x 6 columns]

# Question 8: 

In [64]:
# Total number of rows in the last created dataframe - df_dinner
df_dinner.shape

(7650, 5)

# The df_dinner dataframe has a total of 7650 rows.

# Question 9:

In [80]:
# Information about the customer with a user_id of 1
df_user1 = df_ords[df_ords['user_id'] == '1']

In [83]:
df_user1

order_id user_id  order_number  orders_days_of_week  order_hour_of_day  \
0   2539329       1             1                    2                  8   
1   2398795       1             2                    3                  7   
2    473747       1             3                    3                 12   
3   2254736       1             4                    4                  7   
4    431534       1             5                    4                 15   
5   3367565       1             6                    2                  7   
6    550135       1             7                    1                  9   
7   3108588       1             8                    1                 14   
8   2295261       1             9                    1                 16   
9   2550362       1            10                    4                  8   
10  1187899       1            11                    4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

# Question 10:

In [ ]:
# Details about the user1 behavior - basic stats

In [84]:
# Displays first 5 rows
df_user1.head()

order_id user_id  order_number  orders_days_of_week  order_hour_of_day  \
0  2539329       1             1                    2                  8   
1  2398795       1             2                    3                  7   
2   473747       1             3                    3                 12   
3  2254736       1             4                    4                  7   
4   431534       1             5                    4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [85]:
# Displays last 5 rows
df_user1.tail()

order_id user_id  order_number  orders_days_of_week  order_hour_of_day  \
6    550135       1             7                    1                  9   
7   3108588       1             8                    1                 14   
8   2295261       1             9                    1                 16   
9   2550362       1            10                    4                  8   
10  1187899       1            11                    4                  8   

    days_since_prior_order  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

In [86]:
# Displays total number of rows and columns
df_user1.shape

(11, 6)

In [88]:
# Displays information about the dataframe
df_user1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 11 entries, 0 to 10
Data columns (total 6 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   order_id                11 non-null     object 
 1   user_id                 11 non-null     object 
 2   order_number            11 non-null     int64  
 3   orders_days_of_week     11 non-null     int64  
 4   order_hour_of_day       11 non-null     int64  
 5   days_since_prior_order  10 non-null     float64
dtypes: float64(1), int64(3), object(2)
memory usage: 616.0+ bytes


In [89]:
# Displays the data types of the variables in the dataframe
df_user1.dtypes

order_id                   object
user_id                    object
order_number                int64
orders_days_of_week         int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [90]:
# Displays the columns in the dataframe
df_user1.columns

Index(['order_id', 'user_id', 'order_number', 'orders_days_of_week',
       'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')

In [91]:
# Displays all the descriptive statistics
df_user1.describe()

order_number  orders_days_of_week  order_hour_of_day  \
count     11.000000            11.000000          11.000000   
mean       6.000000             2.636364          10.090909   
std        3.316625             1.286291           3.477198   
min        1.000000             1.000000           7.000000   
25%        3.500000             1.500000           7.500000   
50%        6.000000             3.000000           8.000000   
75%        8.500000             4.000000          13.000000   
max       11.000000             4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

# Question 12 & 13: Exporting data

In [107]:
# Exporting df_ords dataframe
df_ords.to_csv(os.path.join(path, 'Data', 'Prepared Data', 'orders_wrangled.csv'))

In [109]:
# Exporting df_dep_t_new dataframe
df_dep_t_new.to_csv(os.path.join(path, 'Data', 'Prepared Data', 'departments_wrangled.csv')) 